### PR TITLE
Add support for ip and x-header based login

### PR DESCRIPTION
**Pull request by ubergeek42**
This adds 2 new authentication methods, configured by `parameters.yml`.
There's also some commented lines added to app.php and app_dev.php about setting trusted proxies so the ip address seen by symfony will be correct.

## IP Address
On first login, if ipaddress method is enabled, it will set the ip address of a user to whatever ip it found. When viewing the login page it will show you all users that have an ip that matches yours, and allow you to select the user to log in to.

## X-Headers
This method is used by the SER regional, they have a proxy running locally on the team machines that injects a header into the request. This prevents teams from knowing their own login information, and doesn't require relying on the ip address.
This method looks for 3 headers to be set. `X-DOMJUDGE-AUTOLOGIN` must be set to something or this method is ignored. Then it reads the username from `X-DOMJUDGE-LOGIN` and password from `X-DOMJUDGE-PASS`.

This is what it looks like when there are two users that match your ip address.
image

### CODE REVIEW

File: webapp/src/DOMJudgeBundle/Security/Authentication/Token/AlternateLoginToken.php [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        $this->setAuthenticated(count($roles) > 0);
+    }
+
+    public function getCredentials()
+    {
+        return '';
+    }
+
+    public function getIpAddress() {
+      return $this->ipaddress;
```
Comment (nickygerritsen) [inline webapp/src/DOMJudgeBundle/Security/Authentication/Token/AlternateLoginToken.php]: Formatting here seems to be not correct

**Comment by nickygerritsen** [comment]
For the NWERC I got IP auth working for the API as well using this: https://github.com/GEHACK/domjudge/commit/e4510c9239eace534b79c300aafdec98fb51af18

However our approach was different as it doesn't use a button you need to click on. But this might be useful :)

File: webapp/src/DOMJudgeBundle/Controller/SecurityController.php [hunk 2]
```diff
@@ -29,12 +29,12 @@ public function __construct(DOMJudgeService $DOMJudgeService)
      */
     public function loginAction(Request $request)
     {
+        $clientIP = $this->container->get('request_stack')->getMasterRequest()->getClientIp();
         if ($this->get('security.authorization_checker')->isGranted('IS_AUTHENTICATED_FULLY')) {
             $user = $this->get('security.token_storage')->getToken()->getUser();
             $user->setLastLogin(Utils::now());
-            $user->setLastIpAddress(@$_SERVER['REMOTE_ADDR']);
+            $user->setLastIpAddress($this->DOMJudgeService->getClientIp());
```
[the remark this answers is not in the document]
Comment (ubergeek42) [reply webapp/src/DOMJudgeBundle/Controller/SecurityController.php]: Yeah, or rather the domjudgeservice bit should be clientip, and then this should be clientip.

**Comment by ubergeek42** [comment]
New screenshots:
domjudge k8s ubergeek42 com_login
domjudge k8s ubergeek42 com_login 1
domjudge k8s ubergeek42 com_login 2
domjudge k8s ubergeek42 com_login 3
domjudge k8s ubergeek42 com_login 4

I'll update the admin manual, and then merge this.

If you don't update parameters.yaml to include them, the ipaddress and xheaders methods do nothing, so the only one that really matters is the basic auth one. I tried to make it's behavior match what `http_basic: ~` does by default(I couldn't use it because it conflicted with the ipaddress one in some scenarios).

For the ipaddress method, you are able to choose who to log in as when accessing the api, you just need to specify the username(and anything for a password) as basic auth parameters.

**Pull request merged**